src/librustc_typeck/check/mod.rs

```diff
@@ -1709,6 +1717,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     }
 
     pub fn expr_ty(&self, ex: &hir::Expr) -> Ty<'tcx> {
+        if let Some(&adjustment::AdjustNeverToAny(ref t))
+                = self.tables.borrow().adjustments.get(&ex.id) {
+            return t;
+        }
         match self.tables.borrow().node_types.get(&ex.id) {
```

Comment:
I feel like `fcx.node_ty(expr.id)` and `fcx.expr_ty(expr)` are basically equivalent, and suspect there are bits of code that work over both patterns/expressions and just carry an `id` (and they would see the type as `!` and not the type variable). Though ... hmm .. at least `regionck` then goes and adjusts the type, which would be bad.

I wonder if we can rename `expr_ty` to `expr_ty_adjusted` and just have it apply adjustments?